packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -11,7 +11,7 @@ import 'package:flutter/foundation.dart';
 import 'package:flutter/services.dart';
 
 import '_network_image_io.dart'
-  if (dart.library.html) '_network_image_web.dart' as network_image;
+  if (dart.library.js_util) '_network_image_web.dart' as network_image;
```

Comment:
This is a little suspicious to me; even though this uses package:js and package:js/js_util, this line "works" because dart:js_util is available? Can that change in the future? Will this conditional import break if dart:js_util (unused) stops being bundled with flutter web apps?

Reply:
Sorry, this is a bit unclear but classes annotated with `@JS` + `@staticInterop` have their methods lowered to `js_util` calls, so there is an actual dependency even if it is implicit.

This definitely is not the most intuitive way to do conditional compilation though. In any case, it's really our only option for detecting which backend we are on in a conditional import because `dart:html` and `dart:js` aren't supported on Dart2wasm.

Thoughts?

Reply:
> so there is an actual dependency even if it is implicit.

@joshualitt are these two the same?

* https://pub.dev/documentation/js/latest/js_util/js_util-library.html (this is what we have a dependency on, `package:js`, and what is being imported in the js-interop files)
* https://api.dart.dev/stable/2.18.1/dart-js_util/dart-js_util-library.html (this is what drives the conditional imports `dart:js_util`, IIRC)

(Maybe related issue: https://github.com/flutter/flutter/issues/39998)

> our only option for detecting which backend we are on in a conditional import because `dart:html` and `dart:js` aren't supported on Dart2wasm.

Maybe `dart:io`? Do you know if there's any proposal to improve conditional imports to distinguish web vs vm? (I'm asking in case that `dart:js_util` stops being supported by Dart2Wasm at some point, like `dart:html`/`dart:js`?

Reply:
(It seems that the dart SDK can define "_empty" libraries as needed?

https://github.com/flutter/engine/blob/4071af47cfb19d8ef63ecb443420ddb6dd37bea7/sky/packages/sky_engine/lib/_embedder.yaml#L27

(Maybe that can be leveraged to have a more robust detection that is not based on a library that might stop working down the line? (even though it's everywhere in flutter/engine, so we're probably fine?))

Reply:
Yea, I agree in the longer term we want something better here.

I don't think there are any currently in flight proposals to improve the status quo, but I agree we should think about a solution in this space.

In the short term, this is the way we are recommending users detect web. vs native. If any issues occur as a result of this, we will definitely prioritize a replacement solution. The good news is a failure here would be an instant build time issue, so it would be easy to detect and roll back.

Are we okay starting with using the presence of `dart.library.js_util` to detect web vs. native for now, and then revisiting if we ever need to change this invariant?

Reply:
Are we okay starting with using the presence of `dart.library.js_util` to detect web vs. native for now, and then revisiting if we ever need to change this invariant?

I'm fine with this. QQ: why add a dependency on `package:js` to use `js/js_util`, if we already have `dart:js_util` available? Aren't those the same?

Reply:
That is a great question. I'm really not sure why they can't all be in one `dart:` library. Maybe someday?